lib/Parse/ParseDecl.cpp

```diff
@@ -2252,6 +2252,95 @@ static bool parseDifferentiableAttributeArgument(Parser &P,
   return false;
 }
 
+/// Parse the inside of a convention attribute '(...)'.
+///
+/// The '@convention' prefix should've been parsed by the caller.
+/// See `Parser::parseTypeAttribute` for the justChecking argument.
+///
+/// Returns true if there was an error.
+bool Parser::parseConventionAttributeInternal(
+    bool justChecking, TypeAttributes::Convention &convention) {
+  SourceLoc LPLoc;
+  if (!consumeIfNotAtStartOfLine(tok::l_paren)) {
+    if (!justChecking)
+      diagnose(Tok, diag::convention_attribute_expected_lparen);
+    return true;
+  }
+
+  if (Tok.isNot(tok::identifier)) {
+    if (!justChecking)
+      diagnose(Tok, diag::convention_attribute_expected_name);
+    return true;
+  }
+
+  convention.Name = Tok.getText();
+  consumeToken(tok::identifier);
+
+  // TODO: (Varun) The parser should probably not know about what conventions
+  // might have Clang types :(.
```

Comment:
I don't think it's unreasonable, but you can certainly parse it unconditionally if you want and then diagnose if it's provided multiple times.

Reply:
I don't understand what you mean by "provided multiple times". The change I had in mind here was to get rid of the `convention.Name` checks.

Reply:
If you're going to parse generically, you should probably accept an arbitrary list of `<identifier> : <expression>` pairs and then complain about the ones you see multiple times or which you don't understand.  I guess you don't have to, though.

Reply:
Oh, but one advantage of diagnosing this here rather than in Sema is that you can easily suppress follow-on diagnostics from Sema about a missing `cType:` clause by just dropping the attribute entirely.

Reply:
> If you're going to parse generically

That is something I was trying to avoid. We can do that when we need it, instead of right now.

> you can easily suppress follow-on diagnostics from Sema about a missing cType: clause by just dropping the attribute entirely

Not sure why we would diagnose if `cType:` is missing. We don't expect users to write the `cType:` -- would this be a diagnostic specific to swiftinterfaces?

Reply:
I think we should at least require it in SIL types.

Reply:
Ok... but isn't the parser the wrong place to be doing a "error only if the convention is this or that and we are parsing SIL"? It seems convenient but...

Reply:
That should definitely be delayed until Sema.